doc/stubby.1.in
.TH STUBBY 1

.SH NAME
stubby \- a local DNS Privacy stub resolver

.SH SYNOPSIS
.B stubby
[\fB\-C\fR \fIfile\fR]
[\fB\-ghilV\fR]
[\fB\-v\fR \fIloglevel\fR]

.SH DESCRIPTION
.B stubby
acts as a local DNS Privacy stub resolver, using
DNS\-over\-TLS. Stubby encrypts DNS queries sent from the local
machine to a DNS Privacy resolver, increasing end user privacy.
.PP
.B stubby
is in the early stages of development but is suitable for
technical/advanced users.
.PP
.B stubby
provides DNS Privacy by running as a daemon that listens on the
loopback address for DNS queries, and forwards those queries out over TLS.
The default configuration provides Strict Privacy, and uses a subset of
available DNS Privacy servers. See
.I https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Test+Servers
\.

.SH OPTIONS
.TP
.BR \-C \ \fIfile\fR
Read settings from the configuration file \fIfile\fR. If this option is not
given, \fBstubby\fR looks for a configuration file at \fI~/.stubby.yml\fR.
If this is not present, \fBstubby\fR falls back to the global configuration
file \fI@ETCDIR@/stubby.yml\fR.
.TP
.BR \-g
Run \fBstubby\fR as a daemon.
.TP
.BR \-h
Print a usage message and exit.
.TP
.BR \-i
Read the configuration, validate the contents, pretty-print them to
the standard output and exit.
.TP
.BR \-l
Enable all logging. Equivalent to \fB\-v 7\fR.
.TP
.BR \-v \ \fIloglevel\fR
Enable logging. All logging messages at or below \fIloglevel\fR are printed
to standard error:
.RS
.TP
0: EMERG
The system is unusable
.TP
1: ALERT
Action must be taken immediately
.TP
2: CRIT
Critical conditions
.TP
3: ERROR
Error conditions
.TP
4: WARN
Warning conditions
.TP
5: NOTICE
Normal, but significant, conditions
.TP
6: INFO
Information messages
.TP
7: DEBUG
Debug-level messages
.RE
.TP
.BR \-V
Print the \fBstubby\fR version and exit.

.SH CONFIGURATION
The configuration file is in YAML. An example config is:
.PP
.nf
.RS
resolution_type: GETDNS_RESOLUTION_STUB
dns_transport_list:
  - GETDNS_TRANSPORT_TLS
tls_authentication: GETDNS_AUTHENTICATION_REQUIRED
tls_query_padding_blocksize: 256
edns_client_subnet_private : 1
idle_timeout: 10000
listen_addresses:
  - 127.0.0.1
  -  0::1
round_robin_upstreams: 1
upstream_recursive_servers:
  - address_data: 2a04:b900:0:100::38
    tls_auth_name: "getdnsapi.net"
    tls_pubkey_pinset:
      - digest: "sha256"
        value: foxZRnIh9gZpWnl+zEiKa0EJ2rdCGroMWm02gaxSc9Q=
.RE
.fi
.PP
The configuration items are as follow. Some take constant values from
the
.B getdns
library underlying
.B
stubby
and further explanation of their values may be found in the
.B getdns
documentation or at
.I http://www.getdnsapi.net
\.
.TP
.B resolution_type \fIresolution\fR.
This must be set to \fIGETDNS_RESOLUTION_STUB\fR for historic reasons.
.B stubby
will exit with an error if any other setting is used.
.TP
.B dns_transport_list \fIlist\fR
Set the list of transport types to be used. For DNS Privacy this must be set
to \fIGETDNS_TRANSPORT_TLS\fR. Clear text transports are
.I GETDNS_TRANSPORT_TCP
and
.I GETDNS_TRANSPORT_UDP
\.
.TP
.B tls_authentication \fItype\fR
Set the type of authentication required. For Strict Privacy, this
should be set to \fIGETDNS_AUTHENTICATION_REQUIRED\fR. For
Opportunistic mode, remove this setting or set to
\fIGETDNS_AUTHENTICATION_NONE\fR.  In Opportunistic mode
authentication of the nameserver is not required and fallback to clear
text transports is permitted if they are in
.I dns_transport_list
\.
.TP
.B tls_query_padding_blocksize \fIblocksize\fR
Use the EDNS0 padding option to pad DNS queries to hide their size.
.TP
.B edns_client_subnet_private \fI0\fR or \fI1\fR
If 1, use EDNS0 Client Subnet privacy so the client s
ubnet is not sent to authoritative servers.
.TP
.B idle_timeout \fItimeout\fR
Use an EDNS0 Keepalive idle timeout of
.I timeout
milliseconds unless overridden by the server. This keeps idle TLS
connections open to avoid the overhead of opening a new connection
for every query.
.TP
.B round_robin_upstreams \fI0\fR or \fI1\fR
If 1, round robin queries across all the configured upstream
servers. Without this option
.B stubby
will use each upstream server sequentially until it becomes
unavailable and then move on to use the next.
.TP
.B upstream_recursive_servers \fIserver list\fR
Specify the upstream servers that
.B stubby
is to use. Each item in the list contains the following items:
.RS
.TP
.B address_data \fIaddress\fR
IPv4 or IPv6 address of the server.
.TP
.B tls_auth_name \fIname\fR
This is the authentication domain name that will be verified against
the presented certificate.
.TP
.B tls_pubkey_pinset \fIpinset\fR
The sha256 SPKI pinset for the server. This is also verified against
the presented certificate. This contains two items:
.RS
.TP
.B digest \fItype\fR
The type of the key digest.
.TP
.B value \fIkeyval\fR
The key value.
.RE
.RE

.SH FILES
.nf
.I ~/.stubby.yml
.I @ETCDIR@/stubby.yml
.fi

.SH SEE ALSO
.nf
.I https://getdnsapi.net/
.I https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Daemon+-+Stubby
.fi


ChangeLog
*2203-07-26: vesrion 0.4.4-rc1
  * Fix issues with Windows build via PR £335 (thanks AstralStorm)
  * Remove `dnsovertls` servers from the man page
  * Fix broken wiki links (thanks Ayushka Partohap)

* 2022-12-22: version 0.4.3
  * Fix Issue #330 and PR#324: PrivateUsers=false needed in systemd
    stubby.service file for stubby to start.
    Thanks Archcan and Petr Menšík
  * PR #323: Reduce log messages when interface is offline.
    Thanks Russ Bubley and Andre Heider 

* 2022-08-19: version 0.4.2
  * Fix Issue #320: Stubby doesn't start without "log_level"
    entry. Thanks Johnny S. Lee

* 2022-08-19: Version 0.4.1
  * Several updates to the servers in the config file:
     * sinodun.dnsovertls*.com servers are removed and will be
       decommissioned in the near future. This leaves only the
       getdnsapi.net server as the default.
       A recommendation is made that users choose additional servers
       from the list available.
     * Additional Quad9 servers added (thanks pataquets).
     * LDN servers removed as the service is now stopped. 
     * Tidy up of remaining server data.
     * Change `comment` lines on Uncensored server data to more clearly
       be comments (not valid YAML)
     * Fix and improve descriptions of default values
     * getdnsapi.net DoT servers liston on port 443 now too
  * Fixes for parsing of quoted key values for `tls-ca-file`,
    `tls-curves-list` and `resolvconf`
  * Add Dockerfile into `contrib` - thanks to pataquets.
  * Strengthen version checking for LibIDN2 during the cmake step so
    LibIDN2 version 2.3.3 can be properly found (thanks jpbion).
  * Updated systemd/stubby.service file. Thanks Bruno Pagani
  * log level configurable in config file with log_level setting
    Command line setting (with -v or -l) overrules the config file
    setting

* 2021-06-04: Version 0.4.0
  * We announce the intention to remove the dnsovertls*.sinodun.com 
    servers from the default resolver list in the next release,
    see https://github.com/getdnsapi/stubby/issues/286 for details.
  * Add ability to run Stubby as a full Windows service
  * Update IPs for dot1.applied-privacy.net
  * Remove the secure.eu entries in the stubby config file
  * Bugfix: Make `run_in_foreground = 1` the default again
  * Support yaml config files that do not use quotes

* 2020-02-28: Version 0.3.0
  * Correct the applied-privacy.net details and remove ibksturm server in stubby.yml.example.
  * Include AppArmor profile (thanks CameronNemo)
  * Add `contrib` contributors to the README
  * Let scheduled task run indefinitely on Windows (thanks triatic)
  * Add `Restart=on-failure` to the systemd.service file (thanks appliedprivacy)

* 2019-12-20: Version 0.3.0-beta.1
  * Migration of build system to cmake.

* 2019-04-03: Version 0.2.6
 * Windows: use appropriate system and user configuration directories.
 * Windows: replace references to C:\Program Files with %PROGRAMFILES%.
 * Windows: use location of stubby.bat to find stubby.exe and stubby.yml.

* 2019-01-11: Version 0.2.5
 * Fix builds on Windows.
 * RFE getdnsapi/getdns#408: Document trust_anchors_backoff_time
   in stubby.yml.example. Thanks Jonathan Underwood
 * RFE #148: Document tls_ciphersuites, tls_cipher_list, tls_min_version
   and tls_max_version in stubby.yml.example. Thanks Jonathan Underwood
 * RFE #149: Added Google Public DNS to stubby.yml.example.
   Thanks Bruno Pagani

* 2018-12-21: Version 0.2.4
 * DNSSEC required with dnssec extension in example config
 * Removed the yeti servers from stubby.yml.example
 * Added the Foundation RESTENA servers in stubby.yml.example
 * Bugfix: only start Stubby when network is up
   Thanks Bruno Pagani

* 2018-05-11: Version 0.2.3
 * Bugfix #62 and #106: With systemd setups, make /run/stubby directory
   writeable for stubby user and include a "appdata_dir" directory
   in stubby.yml.example.
   Thanks Paul Wouters, eccgecko and Han Vinke
 * Update securedns.eu entries in stubby.yml.example
 * Added Cloudflare servers in stubby.yml.example
 * Added basic upstart script in contrib/upstart dir. Thanks vapniks
 * Bugfix #98: EDNS options that are handled internally should not
   be passed on through downstream.  Thanks Twisteroid Ambassador

* 2018-02-21: Version 0.2.2
 * Fixes and updates to the stubby.yml.config file. Add separate entries for
   servers that listen on port 443.
 * Additional logging of basic config on startup
 * -V option to show version
 * Added a man page

* 2017-12-18: Version 0.2.1
 * Fix use of logging on macos 10.11

* 2017-12-18: Version 0.2.0
 * Add Powershell scripts for Windows 7 that will update the IPv4 DNS resolvers.
 * Add Windows scripts to enable a Scheduled task for stubby
 * Add files to support a separate macOS GUI application to manage stubby
   https://dnsprivacy.org/wiki/x/CIBn
 * Add Quad9 details to the configuration file
 * Bugfix #48: Do not do native DNSSEC validation when cd bit was
   received (for example from an unbound forwarder), but DNSSEC
   validation was not enabled in the first place.
 * Bugfix getdnsapi/getdns#358: Parse config files given with the
   -C option that have an .yaml extension as YAML not JSON.
   Thanks Ollivier Robert

* 2017-11-03: Version 0.1.5
 * Add Windows installer package. Installer available at dnsprivacy.org
 * Fix to systemd file names (thanks ArchangeGabriel)
 * Add SPKI for Uncensored DNS (thanks woopstar)
 * Fix installation of stubby.yml file (thanks ArchangeGabriel)
 * Fix detection of platform for standalone build
 * Fix location of pid file installation
 * Update the stubby.yml file to contain details of all available servers. Only
   a small subset are enabled by default.

* 2017-10-20: Version 0.1.4
 *  '-i' option of stubby no longer tries to bind to the listen addresses so
    it can be run without requiring root privileges. Makes it easier to validate
    the configuration file syntax. 
 *  Fix incorrect IP addresses for some servers in the config file. Add note that

src/stubby.c
/*
 * Copyright (c) 2013, NLNet Labs, Verisign, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * * Neither the names of the copyright holders nor the
 *   names of its contributors may be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL Verisign, Inc. BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include <getdns/getdns.h>
#include <getdns/getdns_extra.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <limits.h>
#include <signal.h>
#include <limits.h>
#ifndef HAVE_GETOPT
#include "getopt.h"
#else
#include <unistd.h>
#endif
#if defined(ENABLE_SYSTEMD)
#include <systemd/sd-daemon.h>
#endif

#include "configfile.h"
#include "log.h"
#include "server.h"
#include "util.h"

#if defined(ENABLE_WINDOWS_SERVICE)
#include "windowsservice.h"
#else
#define STUBBYPIDFILE RUNSTATEDIR"/stubby.pid"
#endif

void
print_usage(FILE *out)
{
	char *home_conf_fn = home_config_file();
	char *system_conf_fn = system_config_file();
	fprintf(out, "usage: " STUBBY_PACKAGE " [<option> ...] \\\n");
	fprintf(out, "\t-C\t<filename>\n");
	fprintf(out, "\t\tRead settings from config file <filename>\n");
	fprintf(out, "\t\tThe getdns context will be configured with these settings\n");
	fprintf(out, "\t\tThe file should be in YAML format with an extension of .yml.\n");
	fprintf(out, "\t\t(The old JSON dict format (.conf) is also still supported when\n");
	fprintf(out, "\t\tspecified on the command line.)\n");
	fprintf(out, "\t\tBy default, the configuration file location is obtained\n");
	fprintf(out, "\t\tby looking for YAML files in the following order:\n");
	fprintf(out, "\t\t\t\"%s\"\n", home_conf_fn);
	fprintf(out, "\t\t\t\"%s\"\n", system_conf_fn);
	fprintf(out, "\t\tA default file (Using Strict mode) is installed as\n");
	fprintf(out, "\t\t\t\"%s\"\n", system_conf_fn);
#if !defined(STUBBY_ON_WINDOWS)
	fprintf(out, "\t-g\tRun stubby in background (default is foreground)\n");
#endif
	fprintf(out, "\t-h\tPrint this help\n");
	fprintf(out, "\t-i\tValidate and print the configuration only. Useful to validate config file\n");
	fprintf(out, "\t\tcontents. Note: does not attempt to bind to the listen addresses.\n");
	fprintf(out, "\t-l\tEnable logging of all logs (same as -v 7)\n");
	fprintf(out, "\t-v\tSpecify logging level (overrides -l option). Values are\n");
	fprintf(out, "\t\t\t0: EMERG  - %s\n", GETDNS_LOG_EMERG_TEXT);
	fprintf(out, "\t\t\t1: ALERT  - %s\n", GETDNS_LOG_ALERT_TEXT);
	fprintf(out, "\t\t\t2: CRIT   - %s\n", GETDNS_LOG_CRIT_TEXT);
	fprintf(out, "\t\t\t3: ERROR  - %s\n", GETDNS_LOG_ERR_TEXT);
	fprintf(out, "\t\t\t4: WARN   - %s\n", GETDNS_LOG_WARNING_TEXT);
	fprintf(out, "\t\t\t5: NOTICE - %s\n", GETDNS_LOG_NOTICE_TEXT);
	fprintf(out, "\t\t\t6: INFO   - %s\n", GETDNS_LOG_INFO_TEXT);
	fprintf(out, "\t\t\t7: DEBUG  - %s\n", GETDNS_LOG_DEBUG_TEXT);
	fprintf(out, "\t-V\tPrint the " STUBBY_PACKAGE " version\n");
	free(home_conf_fn);
	free(system_conf_fn);
}

void
print_version(FILE *out)
{
	fprintf(out, STUBBY_PACKAGE_STRING "\n");
}

extern long log_level;
#define NO_LOGGING 999
int
main(int argc, char **argv)
{
	const char *custom_config_fn = NULL;
	int print_api_info = 0;
	int run_in_foreground = 1;
	int log_connections = 0;
	int dnssec_validation = 0;
#if defined(ENABLE_WINDOWS_SERVICE)
	int windows_service = 0;
	const char *windows_service_arg = NULL;
#endif
	getdns_context  *context = NULL;
	getdns_return_t r;
	getdns_return_t previous_schedule_r = GETDNS_RETURN_GOOD;
	int opt;
	long config_log_level = NO_LOGGING;
	char *ep;

	while ((opt = getopt(argc, argv, "C:ighlv:w:V")) != -1) {
		switch (opt) {
		case 'C':
			custom_config_fn = optarg;
			break;
		case 'g':
			run_in_foreground = 0;
			break;
		case 'h':
			print_usage(stdout);
			exit(EXIT_SUCCESS);
		case 'i':
			print_api_info = 1;
			break;
		case 'l':
			log_connections = 1;
			break;
		case 'v':
			log_connections = 1;
			errno = 0;
			log_level = strtol(optarg, &ep, 10);
			if (log_level < 0 ||  log_level > 7 || *ep != '\0' || 
			    (errno == ERANGE &&
			    (log_level == LONG_MAX || log_level == LONG_MIN)) ) {
				stubby_error("Log level '%s' is invalid or out of range (0-7)", optarg);
				exit(EXIT_FAILURE);
			}
			break;
#if defined(ENABLE_WINDOWS_SERVICE)
		case 'w':
			windows_service = 1;
			windows_service_arg = optarg;
			break;
#endif
                case 'V':
			print_version(stdout);
			exit(EXIT_SUCCESS);
		default:
			print_usage(stderr);
			exit(EXIT_FAILURE);
		}
	}
#if defined(ENABLE_WINDOWS_SERVICE)
	if ( windows_service ) {
		windows_service_command(windows_service_arg, log_connections ? log_level : 0, custom_config_fn);
		exit(EXIT_SUCCESS);
	}
#endif

	if ((r = getdns_context_create(&context, 1))) {
		stubby_error("Create context failed: %s",
		        stubby_getdns_strerror(r));
		return r;
	}
	if (log_connections)
		stubby_set_getdns_logging(context, (int)log_level);

	init_config(context, &config_log_level);
	if ( !read_config(context, custom_config_fn, &dnssec_validation, &config_log_level) )
		exit(EXIT_FAILURE);
	if (config_log_level != NO_LOGGING && !log_connections) {
		log_connections = 1;
		log_level = config_log_level;
	}
	stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_INFO,
		   "Stubby version: %s\n", STUBBY_PACKAGE_STRING);

	if (print_api_info) {
		char *api_information_str = config_get_api_info(context);
		fprintf(stdout, "%s\n", api_information_str);
		free(api_information_str);
		fprintf(stderr, "Result: Config file syntax is valid.\n");
		r = EXIT_SUCCESS;
		goto tidy_and_exit;
	}

	if ( !server_listen(context, (void *)&previous_schedule_r, dnssec_validation)) {
		r = EXIT_FAILURE;
		goto tidy_and_exit;
	}
	
#if !defined(STUBBY_ON_WINDOWS)
	if (!run_in_foreground) {
		pid_t pid;
		char pid_str[1024], *endptr;
		FILE *fh = fopen(STUBBYPIDFILE, "r");
		do {
			pid_t running;

			if (!fh || !fgets(pid_str, sizeof(pid_str), fh))
				break;

			running = strtol(pid_str, &endptr, 10);
			if (endptr == pid_str)
				break;

			if (kill(running, 0) < 0 && errno == ESRCH)
				break;

			stubby_error("Not starting because a running "
				     "stubby was found on pid: %d", running);
			exit(EXIT_FAILURE);
		} while(0);
		if (fh)
			(void) fclose(fh);

		pid = fork();
		if (pid == -1) {
			perror("Could not fork of stubby daemon\n");
			r = GETDNS_RETURN_GENERIC_ERROR;

		} else if (pid) {
			fh = fopen(STUBBYPIDFILE, "w");
			if (fh) {
				fprintf(fh, "%d", (int)pid);
				fclose(fh);
			} else {
				stubby_error("Could not write pid to "
					     "\"%s\": %s", STUBBYPIDFILE,
					     strerror(errno));
				exit(EXIT_FAILURE);
			}
		} else {
#ifdef SIGPIPE
			(void)signal(SIGPIPE, SIG_IGN);
#endif
#ifdef ENABLE_SYSTEMD
			sd_notifyf(0, "READY=1\nMAINPID=%u", getpid());
#endif
			getdns_context_run(context);
		}
	} else
#endif
	{
		/* Report basic config options which specifically affect privacy and validation*/
		stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_INFO,
			   "DNSSEC Validation is %s\n", dnssec_validation==1 ? "ON":"OFF");
		size_t transport_count = 0;
		getdns_transport_list_t *transport_list;
		getdns_context_get_dns_transport_list(context, 
		                                 &transport_count, &transport_list);
		stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_INFO,
			   "Transport list is:\n");
		for (size_t i = 0; i < transport_count; i++) {
			char* transport_name;
			switch (transport_list[i]) {
				case GETDNS_TRANSPORT_UDP:
					transport_name = "UDP";
					break;
				case GETDNS_TRANSPORT_TCP:
					transport_name = "TCP";
					break;
				case GETDNS_TRANSPORT_TLS:
					transport_name = "TLS";
					break;
				default:
					transport_name = "Unknown transport type";
					break;
				}
			stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_INFO,
			                 "  - %s\n", transport_name);
		}
		free(transport_list);
		getdns_tls_authentication_t auth;
		getdns_context_get_tls_authentication(context, &auth);
		stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_INFO,
			   "Privacy Usage Profile is %s\n",
			   auth==GETDNS_AUTHENTICATION_REQUIRED ?
			   "Strict (Authentication required)":"Opportunistic");
		stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_INFO,
			   "(NOTE a Strict Profile only applies when TLS is the ONLY transport!!)\n");
		stubby_log(NULL,GETDNS_LOG_UPSTREAM_STATS, GETDNS_LOG_DEBUG,
			   "Starting DAEMON....\n");
#ifdef SIGPIPE
		(void)signal(SIGPIPE, SIG_IGN);
#endif
#ifdef ENABLE_SYSTEMD
		sd_notify(0, "READY=1");
#endif
		getdns_context_run(context);
	}

tidy_and_exit:
	getdns_context_destroy(context);

	delete_config();

	return r;
}
